share/eds/store.go

```diff
@@ -0,0 +1,231 @@
+package eds
+
+import (
+	"context"
+	"fmt"
+	"io"
+	"os"
+
+	"github.com/filecoin-project/dagstore"
+	"github.com/filecoin-project/dagstore/index"
+	"github.com/filecoin-project/dagstore/mount"
+	"github.com/filecoin-project/dagstore/shard"
+	"github.com/ipfs/go-datastore"
+
+	"github.com/celestiaorg/celestia-node/share"
+
+	"github.com/celestiaorg/rsmt2d"
+)
+
+const (
+	blocksPath     = "/blocks/"
+	indexPath      = "/index/"
+	transientsPath = "/transients/"
+)
+
+// Store maintains (via DAGStore) a top-level index enabling granular and efficient random access to
+// every share and/or Merkle proof over every registered CARv1 file. The EDSStore provides a custom
+// Blockstore interface implementation to achieve access. The main use-case is randomized sampling
+// over the whole chain of EDS block data and getting data by namespace.
+type Store struct {
+	dgstr  *dagstore.DAGStore
+	mounts *mount.Registry
+
+	topIdx index.Inverted
+	carIdx index.FullIndexRepo
+
+	basepath string
+}
+
+// NewStore creates a new EDS Store under the given basepath and datastore.
+func NewStore(basepath string, ds datastore.Batching) (*Store, error) {
+	err := setupPath(basepath)
+	if err != nil {
+		return nil, fmt.Errorf("failed to setup EDSStore directories: %w", err)
+	}
+
+	r := mount.NewRegistry()
+	err = r.Register("fs", &mount.FSMount{FS: os.DirFS(basepath + blocksPath)})
+	if err != nil {
+		return nil, fmt.Errorf("failed to register FS mount on the registry: %w", err)
+	}
+
+	fsRepo, err := index.NewFSRepo(basepath + indexPath)
+	if err != nil {
+		return nil, fmt.Errorf("failed to create index repository: %w", err)
+	}
+
+	invertedRepo := index.NewInverted(ds)
+	dagStore, err := dagstore.NewDAGStore(
+		dagstore.Config{
+			TransientsDir: basepath + transientsPath,
+			IndexRepo:     fsRepo,
+			Datastore:     ds,
+			MountRegistry: r,
+			TopLevelIndex: invertedRepo,
+		},
+	)
+	if err != nil {
+		return nil, fmt.Errorf("failed to create DAGStore: %w", err)
+	}
+
+	s := &Store{
+		basepath: basepath,
+		dgstr:    dagStore,
+		topIdx:   invertedRepo,
+		carIdx:   fsRepo,
+		mounts:   r,
+	}
+
+	return s, nil
+}
+
+// Start starts the underlying DAGStore.
+func (s *Store) Start(ctx context.Context) error {
+	return s.dgstr.Start(ctx)
+}
+
+// Stop stops the underlying DAGStore.
+func (s *Store) Stop() error {
+	return s.dgstr.Close()
```

Comment:
Indexes shouldn't be stopped or stopped by the dagstore, right?

Reply:
The index interfaces do not have Start/Stop, and a dagstore.Close just finishes ongoing writes and cancels the ctx